### ISSUE TITLE
socket_select_interrupter busy loop when waking up from sleep

### ISSUE THREAD

**Issue by arvidn** [chriskohlhoff/asio#461]
When windows goes to sleep, all TCP connections are force-closed. When the computer wakes up again, all socket handles will have an error and will need to be closed.

The `socket_select_interrupter` does not close and re-open when this happens. In [socket_select_interrupter::reset()](https://github.com/chriskohlhoff/asio/blob/master/asio/include/asio/detail/impl/socket_select_interrupter.ipp#L158), where the read-side is drained, all errors are ignored. Meaning EOF is ignored, and the closed socket handle will then be placed back into the `select()` call, which will cause an immediate wake-up again.

I would expect that detecting an error in `reset()` and call `recreate()` would solve this. I have not yet tested this tough.

**Comment by djarek** [comment]
Only tangentially related: why are you using the select backend in ASIO, rather than IOCP? I assume you forcibly changed this setting, because ASIO defaults to using IOCP on Windows.

**Comment by arvidn** [comment]
No, I'm using the default configuration.

**Comment by arvidn** [comment]
I've rebooted into windows 10 and tested this now. As far as I can tell, this patch seems to fix the issue:

to socket_select_interrupter.ipp:
```
  socket_ops::buf b;
   socket_ops::init_buf(b, data, sizeof(data));
   boost::system::error_code ec;
   int bytes_read = socket_ops::recv(read_descriptor_, &b, 1, 0, ec);
+  if (ec) {
+	  recreate();
+	  return false;
+  }
   bool was_interrupted = (bytes_read > 0);
   while (bytes_read == sizeof(data))
     bytes_read = socket_ops::recv(read_descriptor_, &b, 1, 0, ec);
```

Perhaps returning `true` would be safer, since someone *might* have issued an interrupt, but failed.

**Comment by arvidn** [comment]
for completeness; I didn't put the computer to sleep, I used the `currPorts` tool to manually close those TCP connections.

**Comment by chriskohlhoff** [comment]
Please try 4b37bdbe3ed0830af01e9d79de83632161ad0dd3 (which is on master).

**Comment by chriskohlhoff** [comment]
Addressed in asio version 1.16.1.
